File: Backend/src/main/java/fr/rowlaxx/quizzerai/game/Game.java
package fr.rowlaxx.quizzerai.game;

import com.fasterxml.jackson.annotation.JsonIgnore;
import fr.rowlaxx.quizzerai.player.Player;
import fr.rowlaxx.quizzerai.question.Question;
import lombok.Data;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@Data
public class Game {

    private final int code;
    private final int count;
    private final String theme;
    private final GameDifficulty difficulty;
    private final Player owner;

    private boolean generating = true;
    private boolean started = false;
    private boolean finished = false;
    private List<Player> players = new ArrayList<>();

    private final List<Question> questions = new ArrayList<>();
    private GameState state = GameState.GENERATING;
    private Instant stateSince = Instant.now();
    private int currentQuestionIndex;
    private int countdown;

    @JsonIgnore
    public Question getCurrentQuestion() {
        return questions.get(currentQuestionIndex);
    }

    public void setState(GameState state) {
        this.state = state;
        this.stateSince = Instant.now();
    }

    public Map<UUID, Integer> getPoints() {
        return questions.stream().flatMap(e -> e.getAnswers().entrySet().stream())
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> e.getValue().getPoint(),
                        Integer::sum
                ));
    }
}

File: Frontend/quiz_app/lib/dto/game.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:quiz_app/dto/player.dart';
import 'package:quiz_app/dto/question.dart';

part 'game.freezed.dart';
part 'game.g.dart';

@freezed
class Game with _$Game {

  const factory Game({
    required int code,
    required int count,
    required String theme,
    required String difficulty,
    required Player owner,
    required bool generating,
    required bool started,
    required bool finished,
    required List<Player> players,
    required List<Question> questions,
    required String state,
    required double stateSince,
    required int currentQuestionIndex,
    required int countdown,
    required Map<String, double> points
  }) = _Game;

  factory Game.fromJson(Map<String, Object?> json) => _$GameFromJson(json);

}